client/ext/bigquery/current-timestamp-optimizer.js

```diff
@@ -0,0 +1,186 @@
+     * Returns true if the optimizer should be applied to the widget, otherwise false.
+     *
+     * @param {!DashboardModel} dashboard The active dashboard.
+     * @param {!WidgetModel} widget The widget to apply query optimizations against.
+     * @return {boolean} True if the optimizer should be applied to the current widget,
+     *     otherwise false.
+     */
+    canApply(dashboard, widget) {
+      if (goog.isDefAndNotNull(widget.datasource.config.bigQuery.optimizeCurrentTimestamp.enabled)) {
+        return widget.datasource.config.bigQuery.optimizeCurrentTimestamp.enabled;
+      }
+      
+      return dashboard.config.bigQuery.optimizeCurrentTimestamp.enabled;
+    }
+
+    /**
+     * Returns the effective granularity for a widget on a dashboard.  See the class docstring
+     * for more information on how granularity affects date calculation.
+     *
+     * If the widget .enabled setting is true and the widget's .granularity is specified, then
+     * the widget's .granularity is used.  Otherwise, the dashboard's .granularity is used.
+     * 
+     * @param {!DashboardModel} dashboard The active dashboard.
+     * @param {!WidgetModel} widget The widget to apply query optimizations against.
+     * @return {!CurrentTimestampGranularity} The granularity to apply to the widget's
+     *     CURRENT_TIMESTAMP() functions.
+     */
+    getEffectiveGranularity(dashboard, widget) {
+      let widgetSetting = widget.datasource.config.bigQuery.optimizeCurrentTimestamp;
+      let dashboardSetting = dashboard.config.bigQuery.optimizeCurrentTimestamp;
+
+      if (goog.isDefAndNotNull(widgetSetting.granularity)) {
+        return widgetSetting.granularity;
+      } else {
+        return dashboardSetting.granularity;
+      }
+    }
+
+    /**
+     * Returns the current date.  Used for mocking purposes.
+     *
+     * @return {!Date}
+     */
+    getCurrentDate() {
+      return new Date();
+    }
+
+    /**
+     * Returns the rounded effective date based on the provided granularity.
+     *
+     * @param {!CurrentTimestampGranularity} granularity
+     */
+    getEffectiveDate(granularity) {
+      let ranks = {YEAR: 0, MONTH: 1, DAY: 2, HOUR: 3};
+      let rank = ranks[granularity];
+      goog.asserts.assert(goog.isDefAndNotNull(rank));
+
+      let current = this.getCurrentDate();
+
+      let result = new Date(
+        current.getFullYear(),
+        rank >= ranks.MONTH ? current.getMonth() : 0,
+        rank >= ranks.DAY ? current.getDate() : 1,
+        rank >= ranks.HOUR ? current.getHours() : 0);
+
+      return result;
+    }
+
+    /**
+     * Applies the optimization to the current widget, modifying any CURRENT_TIMESTAMP() instances
+     * in .query_exec.
+     *
+     * @param {!DashboardModel} dashboard The active dashboard.
+     * @param {!WidgetModel} widget The widget to apply query optimizations against.
+     * @param {boolean} force If true, applies the optimization even if canApply returns false.
```

Comment:
This comment is very confusing - why would anyone force applying it if it can't be applied?

Core issue is that "canApply" is IMHO misnamed - it controls if the user wants to use this optimization, not if it's possible. Rename that to 'isEnabled', 'shouldApply' or similar?